File: ios/Sources/MritMesh/Crypto/MeshCrypto.swift
import Foundation
import CryptoKit

/// MeshCrypto — all cryptographic operations for the MRIT mesh (iOS).
///
/// Scheme — IDENTICAL to the Android implementation:
///   Key exchange:   ECDH over P-256
///   Key derivation: SHA-256(rawSharedSecret) → 32-byte AES key
///   Encryption:     AES-256-GCM, 12-byte random nonce, 16-byte auth tag
///
/// Wire format of encrypted payload (binary compatible with Android):
///   [ 12 bytes : nonce ] [ N+16 bytes : ciphertext + auth tag ]
///
/// Both platforms produce and consume the SAME wire format — a packet
/// encrypted on Android can be decrypted on iOS and vice versa.
public enum MeshCrypto {

    // ── Key agreement ──────────────────────────────────────────────────────────

    /// Derive a shared AES-256 symmetric key from our private key and the peer's public key.
    ///
    /// Both nodes independently compute the same key — no key material is ever transmitted.
    ///
    /// Derivation: SHA-256(ECDH_shared_secret) — matches Android exactly.
    public static func computeSharedKey(
        ourPrivateKey:  P256.KeyAgreement.PrivateKey,
        theirPublicKey: P256.KeyAgreement.PublicKey
    ) throws -> SymmetricKey {
        let sharedSecret = try ourPrivateKey.sharedSecretFromKeyAgreement(with: theirPublicKey)

        // Hash the raw ECDH output — matches Android's MessageDigest.getInstance("SHA-256").digest(rawSecret)
        let keyBytes: SymmetricKey = sharedSecret.withUnsafeBytes { secretBytes in
            var hasher = SHA256()
            hasher.update(bufferPointer: secretBytes)
            return SymmetricKey(data: hasher.finalize())
        }
        return keyBytes
    }

    // ── Encryption ─────────────────────────────────────────────────────────────

    /// Encrypt [plaintext] with AES-256-GCM using [key].
    ///
    /// Output format (wire-compatible with Android):
    ///   [12-byte nonce] + [ciphertext] + [16-byte auth tag]
    public static func encrypt(_ plaintext: Data, using key: SymmetricKey) throws -> Data {
        let nonce     = AES.GCM.Nonce()            // random 12-byte nonce
        let sealedBox = try AES.GCM.seal(plaintext, using: key, nonce: nonce)

        // Combine into Android-compatible wire format: nonce + ciphertext + tag
        return Data(nonce) + sealedBox.ciphertext + sealedBox.tag
    }

    // ── Decryption ─────────────────────────────────────────────────────────────

    /// Decrypt data produced by [encrypt] (or by Android's MeshCrypto.encrypt).
    ///
    /// Returns nil if authentication fails (tampered data or wrong key).
    public static func decrypt(_ data: Data, using key: SymmetricKey) -> Data? {
        guard data.count > 28 else { return nil }  // 12-byte nonce + at least 1 byte + 16-byte tag

        let nonceData  = data.prefix(12)
        let remaining  = data.dropFirst(12)
        let tag        = remaining.suffix(16)
        let ciphertext = remaining.dropLast(16)

        return try? {
            let nonce     = try AES.GCM.Nonce(data: nonceData)
            let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
            return try AES.GCM.open(sealedBox, using: key)
        }()
    }

    // ── Hashing ────────────────────────────────────────────────────────────────

    /// Short fingerprint of a packet — used by AckManager to match ACKs to their MSG.
    /// Matches the Android MeshCrypto.packetFingerprint algorithm.
    public static func packetFingerprint(
        srcBytes: Data,
        dstBytes: Data,
        payload:  Data
    ) -> String {
        var hasher = SHA256()
        hasher.update(data: srcBytes)
        hasher.update(data: dstBytes)
        hasher.update(data: payload)
        let digest = hasher.finalize()
        return digest.prefix(4).map { String(format: "%02x", $0) }.joined()
    }
}

File: app/src/main/java/com/mrit/mesh/crypto/MeshCrypto.kt
package com.mrit.mesh.crypto

import android.util.Log
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * MeshCrypto — all cryptographic operations for the MRIT mesh.
 *
 * Scheme:
 *   1. Key exchange  — ECDH (Elliptic Curve Diffie-Hellman) over P-256
 *   2. Key derivation — SHA-256(sharedSecret) → 32-byte AES key
 *   3. Encryption    — AES-256-GCM (authenticated encryption — confidentiality + integrity)
 *
 * Wire format of an encrypted payload:
 *   [ 12 bytes : GCM IV/nonce (random per message) ]
 *   [  N bytes : ciphertext + 16-byte GCM auth tag  ]
 *   Total overhead: 28 bytes per message
 *
 * Why AES-256-GCM:
 *   - Authenticated: any tampering of the ciphertext causes decryption to fail
 *   - Fast: hardware-accelerated on all modern Android devices
 *   - No padding needed (stream cipher mode)
 *   - Natively supported in Android (no external libraries required)
 */
object MeshCrypto {

    private const val TAG          = "MeshCrypto"
    private const val AES_MODE     = "AES/GCM/NoPadding"
    private const val GCM_IV_LEN   = 12    // bytes — NIST recommended for GCM
    private const val GCM_TAG_LEN  = 128   // bits — full 16-byte authentication tag

    // ── Key Agreement ──────────────────────────────────────────────────────────

    /**
     * Derive a shared AES-256 key from our private key and the peer's public key.
     *
     * Both nodes compute the same shared secret independently via ECDH.
     * The raw ECDH output is hashed with SHA-256 to produce a uniform 256-bit AES key.
     *
     * @param ourPrivateKey  This node's EC private key
     * @param theirPublicKey The peer's EC public key (received in their DISCOVER packet)
     * @return AES-256 SecretKey for encrypting/decrypting messages to/from this peer
     */
    fun computeSharedKey(ourPrivateKey: PrivateKey, theirPublicKey: PublicKey): SecretKey {
        val ka = KeyAgreement.getInstance("ECDH")
        ka.init(ourPrivateKey)
        ka.doPhase(theirPublicKey, true)
        val rawSecret = ka.generateSecret()

        // Hash the raw ECDH output — produces a uniformly distributed 256-bit key
        val keyBytes = MessageDigest.getInstance("SHA-256").digest(rawSecret)
        return SecretKeySpec(keyBytes, "AES")
    }

    // ── Encryption ─────────────────────────────────────────────────────────────

    /**
     * Encrypt [plaintext] with AES-256-GCM using [key].
     *
     * A fresh random 12-byte IV is generated for every message — critical for GCM security.
     * IV reuse with the same key would be catastrophic for GCM.
     *
     * @return [IV (12 bytes)] + [ciphertext + auth tag (N+16 bytes)]
     */
    fun encrypt(plaintext: ByteArray, key: SecretKey): ByteArray {
        val iv = ByteArray(GCM_IV_LEN).also { SecureRandom().nextBytes(it) }

        val cipher = Cipher.getInstance(AES_MODE)
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(GCM_TAG_LEN, iv))
        val ciphertext = cipher.doFinal(plaintext)

        // Prepend IV so the receiver can extract it for decryption
        return iv + ciphertext
    }

    // ── Decryption ─────────────────────────────────────────────────────────────

    /**
     * Decrypt data produced by [encrypt].
     *
     * GCM authentication is checked automatically — if the ciphertext or IV has been
     * tampered with, this returns null (AEADBadTagException is caught).
     *
     * @param data  [IV (12 bytes)] + [ciphertext + auth tag]
     * @param key   The shared AES-256 key derived via ECDH
     * @return      Decrypted plaintext, or null if authentication failed
     */
    fun decrypt(data: ByteArray, key: SecretKey): ByteArray? {
        if (data.size < GCM_IV_LEN + GCM_TAG_LEN / 8) {
            Log.w(TAG, "Encrypted payload too short: ${data.size} bytes")
            return null
        }

        return try {
            val iv         = data.copyOfRange(0, GCM_IV_LEN)
            val ciphertext = data.copyOfRange(GCM_IV_LEN, data.size)

            val cipher = Cipher.getInstance(AES_MODE)
            cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(GCM_TAG_LEN, iv))
            cipher.doFinal(ciphertext)
        } catch (e: Exception) {
            // javax.crypto.AEADBadTagException → tampered or wrong key
            Log.w(TAG, "Decryption failed (bad tag or wrong key): ${e.javaClass.simpleName}")
            null
        }
    }

    // ── Hashing ────────────────────────────────────────────────────────────────

    /**
     * Compute a short (8 hex char) fingerprint of a packet's identity.
     * Used by AckManager to match ACK packets to their original MSG.
     *
     * Input: srcId bytes + dstId bytes + payload → SHA-256 → first 4 bytes as hex
     */
    fun packetFingerprint(
        srcBytes: ByteArray,
        dstBytes: ByteArray,
        payload: ByteArray
    ): String {
        val digest = MessageDigest.getInstance("SHA-256")
        digest.update(srcBytes)
        digest.update(dstBytes)
        digest.update(payload)
        return digest.digest().take(4).joinToString("") { "%02x".format(it) }
    }
}
